File: sdk/go/terminal_test.go
package agenttrace

import (
	"context"
	"testing"
	"time"
)

func TestTraceTerminalCmd(t *testing.T) {
	client := New(Config{
		APIKey: "test-api-key",
	})
	defer client.Shutdown()

	ctx := context.Background()
	trace := client.Trace(ctx, TraceOptions{
		Name: "test-trace",
	})

	t.Run("tracks command with minimal options", func(t *testing.T) {
		cmd := trace.TerminalCmd(TerminalCommandOptions{
			Command: "npm",
		})

		if cmd.Command != "npm" {
			t.Errorf("expected command to be 'npm', got '%s'", cmd.Command)
		}
		if cmd.TraceID != trace.id {
			t.Error("expected terminal cmd to have trace's ID")
		}
		if cmd.ID == "" {
			t.Error("expected terminal cmd ID to be generated")
		}
		if len(cmd.Args) != 0 {
			t.Error("expected empty args by default")
		}
		if cmd.WorkingDirectory == "" {
			t.Error("expected working directory to be set")
		}
	})

	t.Run("tracks command with args", func(t *testing.T) {
		cmd := trace.TerminalCmd(TerminalCommandOptions{
			Command: "npm",
			Args:    []string{"install", "--save", "lodash"},
		})

		if cmd.Command != "npm" {
			t.Errorf("expected command to be 'npm', got '%s'", cmd.Command)
		}
		if len(cmd.Args) != 3 {
			t.Errorf("expected 3 args, got %d", len(cmd.Args))
		}
		if cmd.Args[0] != "install" || cmd.Args[1] != "--save" || cmd.Args[2] != "lodash" {
			t.Error("args not set correctly")
		}
	})

	t.Run("tracks command with output", func(t *testing.T) {
		cmd := trace.TerminalCmd(TerminalCommandOptions{
			Command:  "echo",
			Args:     []string{"hello"},
			Stdout:   "hello\n",
			Stderr:   "",
			ExitCode: 0,
		})

		if cmd.Stdout != "hello\n" {
			t.Errorf("expected stdout to be 'hello\\n', got '%s'", cmd.Stdout)
		}
		if cmd.Stderr != "" {
			t.Errorf("expected stderr to be empty, got '%s'", cmd.Stderr)
		}
		if cmd.ExitCode != 0 {
			t.Errorf("expected exitCode to be 0, got %d", cmd.ExitCode)
		}
		if cmd.Success != true {
			t.Error("expected success to be true")
		}
	})

	t.Run("tracks failed command", func(t *testing.T) {
		cmd := trace.TerminalCmd(TerminalCommandOptions{
			Command:  "false",
			ExitCode: 1,
			Stderr:   "Error occurred",
		})

		if cmd.ExitCode != 1 {
			t.Errorf("expected exitCode to be 1, got %d", cmd.ExitCode)
		}
		if cmd.Success != false {
			t.Error("expected success to be false")
		}
		if cmd.Stderr != "Error occurred" {
			t.Error("expected stderr to be set")
		}
	})

	t.Run("tracks with custom working directory", func(t *testing.T) {
		cmd := trace.TerminalCmd(TerminalCommandOptions{
			Command:          "ls",
			WorkingDirectory: "/custom/path",
		})

		if cmd.WorkingDirectory != "/custom/path" {
			t.Errorf("expected workingDirectory to be '/custom/path', got '%s'", cmd.WorkingDirectory)
		}
	})

	t.Run("tracks with observation ID", func(t *testing.T) {
		cmd := trace.TerminalCmd(TerminalCommandOptions{
			Command:       "test",
			ObservationID: "obs-456",
		})

		if cmd.ObservationID != "obs-456" {
			t.Errorf("expected observationID to be 'obs-456', got '%s'", cmd.ObservationID)
		}
	})

	t.Run("calculates duration", func(t *testing.T) {
		startedAt := time.Now().Add(-5 * time.Second)
		completedAt := time.Now()

		cmd := trace.TerminalCmd(TerminalCommandOptions{
			Command:     "long-command",
			StartedAt:   &startedAt,
			CompletedAt: &completedAt,
		})

		if cmd.DurationMs < 4900 || cmd.DurationMs > 5100 {
			t.Errorf("expected duration to be ~5000ms, got %d", cmd.DurationMs)
		}
	})

	t.Run("tracks timed out command", func(t *testing.T) {
		cmd := trace.TerminalCmd(TerminalCommandOptions{
			Command:  "slow-command",
			TimedOut: true,
			Killed:   true,
			ExitCode: -1,
		})

		if cmd.ExitCode != -1 {
			t.Error("expected exitCode to be -1 for killed process")
		}
	})

	t.Run("tracks truncated output", func(t *testing.T) {
		cmd := trace.TerminalCmd(TerminalCommandOptions{
			Command:         "verbose-command",
			Stdout:          "truncated...",
			StdoutTruncated: true,
			Stderr:          "errors...",
			StderrTruncated: true,
		})

		if cmd.Stdout != "truncated..." {
			t.Error("expected truncated stdout")
		}
	})

	t.Run("handles nil args", func(t *testing.T) {
		cmd := trace.TerminalCmd(TerminalCommandOptions{
			Command: "test",
			Args:    nil,
		})

		if cmd.Args == nil {
			t.Error("expected args to be initialized to empty slice")
		}
		if len(cmd.Args) != 0 {
			t.Error("expected args to be empty")
		}
	})
}

func TestTraceRunCmd(t *testing.T) {
	client := New(Config{
		APIKey: "test-api-key",
	})
	defer client.Shutdown()

	ctx := context.Background()
	trace := client.Trace(ctx, TraceOptions{
		Name: "test-trace",
	})

	t.Run("runs echo command", func(t *testing.T) {
		result := trace.RunCmd(ctx, "echo", &RunCommandOptions{
			Args: []string{"hello"},
		})

		if result.ExitCode != 0 {
			t.Errorf("expected exitCode to be 0, got %d", result.ExitCode)
		}
		if result.Stdout != "hello\n" {
			t.Errorf("expected stdout to be 'hello\\n', got '%s'", result.Stdout)
		}
		if result.Info == nil {
			t.Error("expected info to be set")
		}
		if result.Info.Command != "echo" {
			t.Errorf("expected command to be 'echo', got '%s'", result.Info.Command)
		}
	})

	t.Run("handles non-zero exit code", func(t *testing.T) {
		result := trace.RunCmd(ctx, "false", nil)

		if result.ExitCode == 0 {
			t.Error("expected non-zero exit code")
		}
		if result.Info.Success != false {
			t.Error("expected success to be false")
		}
	})

	t.Run("handles nonexistent command", func(t *testing.T) {
		result := trace.RunCmd(ctx, "nonexistent-command-12345", nil)

		if result.ExitCode != -1 {
			t.Errorf("expected exitCode to be -1 for failed command, got %d", result.ExitCode)
		}
	})

	t.Run("respects timeout", func(t *testing.T) {
		ctx, cancel := context.WithTimeout(ctx, 100*time.Millisecond)
		defer cancel()

		result := trace.RunCmd(ctx, "sleep", &RunCommandOptions{
			Args:    []string{"10"},
			Timeout: 100 * time.Millisecond,
		})

		// Command should be killed due to timeout
		if result.ExitCode == 0 {
			t.Error("expected command to fail due to timeout")
		}
	})

	t.Run("uses custom working directory", func(t *testing.T) {
		result := trace.RunCmd(ctx, "pwd", &RunCommandOptions{
			WorkingDirectory: "/tmp",
		})

		if result.ExitCode != 0 {
			t.Errorf("expected exitCode to be 0, got %d", result.ExitCode)
		}
		if result.Info.WorkingDirectory != "/tmp" {
			t.Errorf("expected workingDirectory to be '/tmp', got '%s'", result.Info.WorkingDirectory)
		}
	})

	t.Run("passes environment variables", func(t *testing.T) {
		result := trace.RunCmd(ctx, "sh", &RunCommandOptions{
			Args: []string{"-c", "echo $TEST_VAR"},
			Env:  map[string]string{"TEST_VAR": "test_value"},
		})

		if result.ExitCode != 0 {
			t.Errorf("expected exitCode to be 0, got %d", result.ExitCode)
		}
		if result.Stdout != "test_value\n" {
			t.Errorf("expected stdout to contain env var value, got '%s'", result.Stdout)
		}
	})

	t.Run("truncates long output", func(t *testing.T) {
		result := trace.RunCmd(ctx, "sh", &RunCommandOptions{
			Args:           []string{"-c", "yes | head -1000"},
			MaxOutputBytes: 100,
		})

		if len(result.Stdout) > 100 {
			t.Errorf("expected stdout to be truncated to 100 bytes, got %d", len(result.Stdout))
		}
	})
}

func TestTerminalCmdDisabledClient(t *testing.T) {
	enabled := false
	client := New(Config{
		APIKey:  "test-api-key",
		Enabled: &enabled,
	})
	defer client.Shutdown()

	ctx := context.Background()
	trace := client.Trace(ctx, TraceOptions{
		Name: "test-trace",
	})

	cmd := trace.TerminalCmd(TerminalCommandOptions{
		Command: "npm",
		Args:    []string{"test"},
	})

	// Should still return command info
	if cmd.Command != "npm" {
		t.Error("expected command info even when disabled")
	}
	if cmd.ID == "" {
		t.Error("expected command ID even when disabled")
	}
}

func TestTerminalCommandInfo(t *testing.T) {
	t.Run("has all expected fields", func(t *testing.T) {
		now := time.Now()
		info := TerminalCommandInfo{
			ID:               "cmd-123",
			TraceID:          "trace-456",
			ObservationID:    "obs-789",
			Command:          "npm",
			Args:             []string{"test"},
			WorkingDirectory: "/path/to/project",
			ExitCode:         0,
			Stdout:           "output",
			Stderr:           "",
			Success:          true,
			DurationMs:       1000,
			StartedAt:        now,
			CompletedAt:      now.Add(time.Second),
		}

		if info.ID != "cmd-123" {
			t.Error("ID not set correctly")
		}
		if info.TraceID != "trace-456" {
			t.Error("TraceID not set correctly")
		}
		if info.ObservationID != "obs-789" {
			t.Error("ObservationID not set correctly")
		}
		if info.Command != "npm" {
			t.Error("Command not set correctly")
		}
		if len(info.Args) != 1 || info.Args[0] != "test" {
			t.Error("Args not set correctly")
		}
		if info.WorkingDirectory != "/path/to/project" {
			t.Error("WorkingDirectory not set correctly")
		}
		if info.ExitCode != 0 {
			t.Error("ExitCode not set correctly")
		}
		if info.Stdout != "output" {
			t.Error("Stdout not set correctly")
		}
		if info.Stderr != "" {
			t.Error("Stderr not set correctly")
		}
		if info.Success != true {
			t.Error("Success not set correctly")
		}
		if info.DurationMs != 1000 {
			t.Error("DurationMs not set correctly")
		}
	})
}

func TestRunCommandOptions(t *testing.T) {
	t.Run("has sensible defaults when nil", func(t *testing.T) {
		client := New(Config{
			APIKey: "test-api-key",
		})
		defer client.Shutdown()

		ctx := context.Background()
		trace := client.Trace(ctx, TraceOptions{
			Name: "test-trace",
		})

		// Should not panic with nil options
		result := trace.RunCmd(ctx, "echo", nil)

		if result.Info.Command != "echo" {
			t.Error("expected command to be set")
		}
	})
}

File: sdk/typescript/tests/terminal.test.ts
/**
 * Tests for the AgentTrace TypeScript SDK terminal module.
 */

import { describe, it, expect, vi, beforeEach, afterEach } from "vitest";
import { spawn } from "child_process";
import { EventEmitter } from "events";

// Mock child_process
vi.mock("child_process", () => ({
  spawn: vi.fn(),
}));

// Mock process.cwd
vi.spyOn(process, "cwd").mockReturnValue("/mock/working/dir");

import { TerminalClient, TerminalCommandInfo, runCommand } from "../src/terminal";

describe("TerminalClient", () => {
  let mockClient: any;
  let terminalClient: TerminalClient;

  beforeEach(() => {
    mockClient = {
      enabled: true,
      _addEvent: vi.fn(),
    };
    terminalClient = new TerminalClient(mockClient);
    vi.clearAllMocks();
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  describe("track", () => {
    it("should track a command with minimal options", () => {
      const cmd = terminalClient.track("trace-123", {
        command: "npm",
      });

      expect(cmd.command).toBe("npm");
      expect(cmd.traceId).toBe("trace-123");
      expect(cmd.id).toBeDefined();
      expect(cmd.args).toEqual([]);
      expect(cmd.workingDirectory).toBe("/mock/working/dir");
    });

    it("should track a command with args", () => {
      const cmd = terminalClient.track("trace-123", {
        command: "npm",
        args: ["install", "--save", "lodash"],
      });

      expect(cmd.command).toBe("npm");
      expect(cmd.args).toEqual(["install", "--save", "lodash"]);
    });

    it("should track command with output", () => {
      const cmd = terminalClient.track("trace-123", {
        command: "echo",
        args: ["hello"],
        stdout: "hello\n",
        stderr: "",
        exitCode: 0,
      });

      expect(cmd.stdout).toBe("hello\n");
      expect(cmd.stderr).toBe("");
      expect(cmd.exitCode).toBe(0);
      expect(cmd.success).toBe(true);
    });

    it("should track failed command", () => {
      const cmd = terminalClient.track("trace-123", {
        command: "false",
        exitCode: 1,
        stderr: "Error occurred",
      });

      expect(cmd.exitCode).toBe(1);
      expect(cmd.success).toBe(false);
      expect(cmd.stderr).toBe("Error occurred");
    });

    it("should track with custom working directory", () => {
      const cmd = terminalClient.track("trace-123", {
        command: "ls",
        workingDirectory: "/custom/path",
      });

      expect(cmd.workingDirectory).toBe("/custom/path");
    });

    it("should track with observation ID", () => {
      const cmd = terminalClient.track("trace-123", {
        command: "test",
        observationId: "obs-456",
      });

      expect(cmd.observationId).toBe("obs-456");
    });

    it("should calculate duration", () => {
      const startedAt = new Date("2024-01-01T10:00:00Z");
      const completedAt = new Date("2024-01-01T10:00:05Z");

      const cmd = terminalClient.track("trace-123", {
        command: "long-command",
        startedAt,
        completedAt,
      });

      expect(cmd.durationMs).toBe(5000);
    });

    it("should send event to API when enabled", () => {
      terminalClient.track("trace-123", {
        command: "npm",
        args: ["test"],
        exitCode: 0,
        stdout: "All tests passed",
      });

      expect(mockClient._addEvent).toHaveBeenCalled();
      const eventArg = mockClient._addEvent.mock.calls[0][0];
      expect(eventArg.type).toBe("terminal-command-create");
      expect(eventArg.body.traceId).toBe("trace-123");
      expect(eventArg.body.command).toBe("npm");
      expect(eventArg.body.args).toEqual(["test"]);
    });

    it("should not send event when client is disabled", () => {
      mockClient.enabled = false;

      const cmd = terminalClient.track("trace-123", {
        command: "npm",
        args: ["test"],
      });

      expect(mockClient._addEvent).not.toHaveBeenCalled();
      expect(cmd.command).toBe("npm"); // Still returns info
    });

    it("should track timed out command", () => {
      const cmd = terminalClient.track("trace-123", {
        command: "slow-command",
        timedOut: true,
        killed: true,
        exitCode: -1,
      });

      expect(mockClient._addEvent).toHaveBeenCalled();
      const eventArg = mockClient._addEvent.mock.calls[0][0];
      expect(eventArg.body.timedOut).toBe(true);
      expect(eventArg.body.killed).toBe(true);
    });

    it("should track truncated output", () => {
      const cmd = terminalClient.track("trace-123", {
        command: "verbose-command",
        stdout: "truncated...",
        stdoutTruncated: true,
        stderr: "errors...",
        stderrTruncated: true,
      });

      expect(mockClient._addEvent).toHaveBeenCalled();
      const eventArg = mockClient._addEvent.mock.calls[0][0];
      expect(eventArg.body.stdoutTruncated).toBe(true);
      expect(eventArg.body.stderrTruncated).toBe(true);
    });
  });

  describe("run", () => {
    it("should run a command and track it", async () => {
      // Create mock process
      const mockProc = new EventEmitter() as any;
      mockProc.stdout = new EventEmitter();
      mockProc.stderr = new EventEmitter();
      vi.mocked(spawn).mockReturnValue(mockProc);

      const runPromise = terminalClient.run("trace-123", "echo", {
        args: ["hello"],
      });

      // Simulate stdout
      mockProc.stdout.emit("data", Buffer.from("hello\n"));

      // Simulate process exit
      mockProc.emit("close", 0);

      const result = await runPromise;

      expect(result.exitCode).toBe(0);
      expect(result.stdout).toBe("hello\n");
      expect(result.info).toBeDefined();
      expect(result.info.command).toBe("echo");
    });

    it("should handle command with non-zero exit code", async () => {
      const mockProc = new EventEmitter() as any;
      mockProc.stdout = new EventEmitter();
      mockProc.stderr = new EventEmitter();
      vi.mocked(spawn).mockReturnValue(mockProc);

      const runPromise = terminalClient.run("trace-123", "false");

      mockProc.stderr.emit("data", Buffer.from("Error\n"));
      mockProc.emit("close", 1);

      const result = await runPromise;

      expect(result.exitCode).toBe(1);
      expect(result.stderr).toBe("Error\n");
      expect(result.info.success).toBe(false);
    });

    it("should handle command error", async () => {
      const mockProc = new EventEmitter() as any;
      mockProc.stdout = new EventEmitter();
      mockProc.stderr = new EventEmitter();
      vi.mocked(spawn).mockReturnValue(mockProc);

      const runPromise = terminalClient.run("trace-123", "nonexistent-command");

      mockProc.emit("error", new Error("Command not found"));

      const result = await runPromise;

      expect(result.exitCode).toBe(-1);
      expect(result.stderr).toBe("Command not found");
      expect(result.info.success).toBe(false);
    });

    it("should respect timeout", async () => {
      vi.useFakeTimers();

      const mockProc = new EventEmitter() as any;
      mockProc.stdout = new EventEmitter();
      mockProc.stderr = new EventEmitter();
      mockProc.kill = vi.fn();
      vi.mocked(spawn).mockReturnValue(mockProc);

      const runPromise = terminalClient.run("trace-123", "sleep", {
        args: ["10"],
        timeout: 1000,
      });

      // Fast-forward past timeout
      vi.advanceTimersByTime(1500);

      expect(mockProc.kill).toHaveBeenCalledWith("SIGTERM");

      // Clean up
      mockProc.emit("close", -1);

      vi.useRealTimers();
    });

    it("should truncate long output", async () => {
      const mockProc = new EventEmitter() as any;
      mockProc.stdout = new EventEmitter();
      mockProc.stderr = new EventEmitter();
      vi.mocked(spawn).mockReturnValue(mockProc);

      const runPromise = terminalClient.run("trace-123", "generate-output", {
        maxOutputBytes: 10,
      });

      // Send more data than maxOutputBytes
      mockProc.stdout.emit("data", Buffer.from("0123456789ABCDEF"));
      mockProc.emit("close", 0);

      const result = await runPromise;

      expect(result.stdout.length).toBeLessThanOrEqual(10);
    });
  });
});

describe("TerminalCommandInfo", () => {
  it("should have all expected fields", () => {
    const mockClient = {
      enabled: false,
      _addEvent: vi.fn(),
    };
    const client = new TerminalClient(mockClient);

    const cmd = client.track("trace-123", {
      command: "npm",
      args: ["test"],
      observationId: "obs-456",
      workingDirectory: "/path",
      exitCode: 0,
      stdout: "output",
      stderr: "errors",
    });

    expect(cmd).toHaveProperty("id");
    expect(cmd).toHaveProperty("traceId");
    expect(cmd).toHaveProperty("observationId");
    expect(cmd).toHaveProperty("command");
    expect(cmd).toHaveProperty("args");
    expect(cmd).toHaveProperty("workingDirectory");
    expect(cmd).toHaveProperty("exitCode");
    expect(cmd).toHaveProperty("stdout");
    expect(cmd).toHaveProperty("stderr");
    expect(cmd).toHaveProperty("success");
    expect(cmd).toHaveProperty("durationMs");
    expect(cmd).toHaveProperty("startedAt");
    expect(cmd).toHaveProperty("completedAt");
  });
});

describe("runCommand", () => {
  it("should throw error when no client available", async () => {
    // Mock getClient to return null
    vi.mock("../src/context", () => ({
      getClient: vi.fn().mockReturnValue(null),
      getCurrentTrace: vi.fn().mockReturnValue(null),
    }));

    await expect(runCommand("echo")).rejects.toThrow("No AgentTrace client available");
  });
});
